<a href="https://colab.research.google.com/github/xpertdesh/ml-class21/blob/main/projects/bach_chorales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center>Bach Chorales Project</center>
---
<br>
This project consists of predicting the chords that Bach used based off from a set of notes. Johann Sebastian Bach was a German composer and musician of the Baroque period.<br> 
Fun fact: This is the second time in my life hearing about Bach. First time, was for 326 when Stephen made us read that humungous book GEB.<br><br>
Before looking at the data set. I understand there are 3 columns of actual data, and 2 that wouldn't be used in the project.

### Ok let's get started!


https://raw.githubusercontent.com/zacharski/ml-class/master/data/bach.zip

*   First,  we will download and unzip the file
*   Then, we will import the neccessary libraries in order to take a look at the data
*   Finally, we will explore and analyze the data


In [9]:
! wget https://raw.githubusercontent.com/zacharski/ml-class/master/data/bach.zip

--2021-02-22 02:00:11--  https://raw.githubusercontent.com/zacharski/ml-class/master/data/bach.zip
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 41761 (41K) [application/zip]
Saving to: ‘bach.zip’

bach.zip            100%[===================>]  40.78K  --.-KB/s    in 0.005s  

2021-02-22 02:00:12 (7.76 MB/s) - ‘bach.zip’ saved [41761/41761]



In [10]:
! unzip bach.zip -d bach_chorales

Archive:  bach.zip
  inflating: bach_chorales/bach.csv  


The previous step actually took me quite a while but that's fine.<br>
Ok, with that out of the way. We can now import the neccessary libraries.<br><br><br>

##1. Import Neccessary Libraries
<br><br><br>

In [5]:
#data analysis libraries 
import numpy as np
import pandas as pd

#visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

#ignore warnings
import warnings
warnings.filterwarnings('ignore')

<br><br><br>

##2. Read in and Explore the Data
It's time to read in the data using `pd.readcsv()`, and take a first look at the data using the `describe()` function.<br><br><br>

In [13]:
bach = pd.read_csv('bach_chorales/bach.csv')
bach.describe(include='all')

choral_ID  event_number     C    C#  ...     B  bass        meter chord_label
count       5665   5665.000000  5665  5665  ...  5665  5665  5665.000000        5665
unique        60           NaN     2     2  ...     2    16          NaN         102
top     002908ch           NaN    NO    NO  ...    NO     D          NaN         D_M
freq         207           NaN  3875  4711  ...  3874   689          NaN         503
mean         NaN     53.374404   NaN   NaN  ...   NaN   NaN     3.134863         NaN
std          NaN     37.268208   NaN   NaN  ...   NaN   NaN     1.109710         NaN
min          NaN      1.000000   NaN   NaN  ...   NaN   NaN     1.000000         NaN
25%          NaN     24.000000   NaN   NaN  ...   NaN   NaN     2.000000         NaN
50%          NaN     48.000000   NaN   NaN  ...   NaN   NaN     3.000000         NaN
75%          NaN     75.000000   NaN   NaN  ...   NaN   NaN     4.000000         NaN
max          NaN    207.000000   NaN   NaN  ...   NaN   NaN     5.000000         NaN

[11 rows x 17 columns]

<br><br><br>

#3. Data Analysis
We're going to consider the features in the dataset and how complete they are.
<br><br><br>

In [17]:
#get of a list of the features within the dataset
bach.columns

Index(['choral_ID', 'event_number', 'C', 'C#', 'D', 'D#', 'E', 'F', 'F#', 'G',
       'G#', 'A', 'A#', 'B', 'bass', 'meter', 'chord_label'],
      dtype='object')

In [20]:
#see a sample of the data set to get an idea of the variables
bach.sample(10)

choral_ID  event_number    C   C#    D  ...   A#    B bass meter chord_label
354   000408b_            55   NO   NO   NO  ...   NO  YES    E     5         E_m
1365  001805b_            35  YES   NO  YES  ...   NO   NO    G     3         G_M
4667  014406b_           103   NO  YES   NO  ...  YES   NO   C#     3        C#d6
584   000707b_            61   NO  YES   NO  ...  YES   NO   F#     3         F#M
4453  014007b_           124  YES   NO   NO  ...   NO   NO    C     3         C_m
1911  002908ch           101   NO   NO  YES  ...   NO   NO   F#     5         D_M
1097  001306b_            52  YES   NO   NO  ...  YES   NO    E     3        C_M7
3604  012606b_            29  YES   NO   NO  ...   NO   NO    G     2         A_m
1453  001907ch            12  YES   NO   NO  ...   NO   NO    E     5         C_M
5031  014806b_            58   NO   NO   NO  ...   NO  YES    B     2         B_M

[10 rows x 17 columns]

<br><br>


* **Numerical Features:** Event Number(Discrete), Meter(Discrete)
* **Categorical Features:** Western Scale(Discrete), Bass(Discrete)
* **Label:** Chord Label(Discrete)

<br>Choral Id, and Event Number will not be used for training.
<br><br>
Let's take another look at the dataset, now that we know what kinds of features we're working with.<br><br><br>


In [21]:
#see a summary of the dataset
bach.describe(include='all')

choral_ID  event_number     C    C#  ...     B  bass        meter chord_label
count       5665   5665.000000  5665  5665  ...  5665  5665  5665.000000        5665
unique        60           NaN     2     2  ...     2    16          NaN         102
top     002908ch           NaN    NO    NO  ...    NO     D          NaN         D_M
freq         207           NaN  3875  4711  ...  3874   689          NaN         503
mean         NaN     53.374404   NaN   NaN  ...   NaN   NaN     3.134863         NaN
std          NaN     37.268208   NaN   NaN  ...   NaN   NaN     1.109710         NaN
min          NaN      1.000000   NaN   NaN  ...   NaN   NaN     1.000000         NaN
25%          NaN     24.000000   NaN   NaN  ...   NaN   NaN     2.000000         NaN
50%          NaN     48.000000   NaN   NaN  ...   NaN   NaN     3.000000         NaN
75%          NaN     75.000000   NaN   NaN  ...   NaN   NaN     4.000000         NaN
max          NaN    207.000000   NaN   NaN  ...   NaN   NaN     5.000000         NaN

[11 rows x 17 columns]

<br><br><br>

**Some Observatios:** 

* There are a total of 5,665 chords in our dataset.
* All the data is complete, no missing values in the dataset.
* We're going to have to hot-encode the Western Scale, and Bass columns.
* Choral Id and Event Number columns will get dropped since they won't help the fitting process.
<br><br>

**Some Predictions:** 
* I don't know anything about chords to be specific, boring lol.
* Maybe I'll ask a friend to predictions from someone who knows.
<br><br><br>

# 4. Data Visualization

Done for today!